### DISCUSSION TITLE
Spearbit audit discussion

### DISCUSSION THREAD

**Discussion by MerlinEgalite** [morpho-org/morpho-optimizers#1526] [General]
✅ Ok, will do
❌ No (most of the time, it is exlained in the issue)
🚧 Need further work
<details>
  <summary>Merlin's List</summary>
- 19 🚧
- 20 ❌
- 21 ❌
- 22 ✅
- 23 ✅
- 24 ✅
- 25 ✅
- 26 🚧 (related to 19)
- 27 ✅ (gas opti, not mandatory)
- 28 ✅
- 29 🚧
- 30 ✅
- 31 ❌ (for future version)
- 32 ❌
- 33 ❌ (not necessary)
- 34 ✅ (for the 0 check, not the upper limit)
- 35 ❌
- 36 ❌ (Aave is not doing it)
- 37 ✅
- 38 ✅ (minor gas opti, not mandatory)
- 39 ✅
- 40 ✅
- 41 ❌
- 42 ❌
- 43 ❌
- 44 ❌
- 45 ✅
- 46 ✅
- 47 ✅
- 48 ✅
- 49 ✅
- 50 ✅
- 51 ✅
- 52 ✅
- 53 ✅
- 54 ✅
- 55 ✅
- 56 ✅
- 57 ✅
- 58 ✅
- 59 ✅
- 60 ✅
- 61 ✅
- 62 ✅
- 63 ✅
- 64 ✅
- 65 ✅
- 66 ✅
- 67 ❌ (Acknowledge but need to conduct quantitative research)
- 68 ❌ (won't change anything about it)
- 69 ✅
- 70 ❌
- 71 ❌ (Acknowledge but need to conduct quantitative research)
</details>

**Comment by pakim249CAL** [comment]
<details>
  <summary>Patrick's List</summary>

* 19 🚧
* 20 ✅
* 21 ❌
* 22 ✅
* 23 ❌
* 24 ❌
* 25 ✅
* 26 🚧
* 27 ✅
* 28 🚧
* 29 🚧
* 30 ✅
* 31 ❌
* 32 ❌
* 33 🚧
* 34 ✅
* 35 ❌
* 36 ❌
* 37 ✅
* 38 ✅
* 39 ✅
* 40 ✅
* 41 ❌
* 42 ❌
* 43 ❌
* 44 ❌
* 45 ✅  
* 46 ❌
* 47 ✅  
* 48 ✅  
* 49 ❌
* 50 ✅  (This is in progress)
* 51 ❌
* 52 ✅ 
* 53 ✅ 
* 54 ✅ 
* 55 ❌
* 56 ✅ (We are already discussing this I think)
* 57 ✅
* 58 ✅
* 59 🚧
* 60 🚧
* 61 ❌
* 62 ❌
* 63 ❌
* 64 ❌ 
* 65 ❌ 
* 66 ✅ (This issue is basically multiple issues. I X out 61-65 because there aren't any unique issues to those that I agree with that isn't in this issue.)
* 67 ❌ This seems known already. We cannot allow Morpho to ever reach a state to be liquidated.
* 68 ❌
* 69 ❌ Can be mitigated with proper user warning.
* 70 ❌ 
* 71 ❌ IMO a risk we can accept, as hard withdraw is a fallback in the first place. But I'm not against further study if we want to pursue it.
</details>

**Comment by QGarchery** [comment]
<details>
  <summary>Quentin's List</summary>
  
19. :construction: we need to decide if gas savings are worth the added safety
20. :construction: it requires to introduce changes for specific markets (cc @MerlinEgalite), like in USDC
21. :x: 
22. :white_check_mark: 
23. :white_check_mark: 
24. :x: There is nothing to do in the code ?
25. :white_check_mark: 
26. :construction: linked to 19
27. :white_check_mark: 
28. :x: this issue should be nullified since we found it (not without sweat I might add)
29. :construction: need research to see if the attack scenario of borrowing the entire pool is ever profitable or dangerous for Morpho
30. :white_check_mark: 
31. :x: too complicated for small improvements IMO. We can try to compute how much rewards was gained by the deltas, but I'm pretty sure it's not a lot
32. :construction: not sure how complicated it is to implement
33. :white_check_mark: 
34. :x: upper value we don't want to do (we may even put `type(uint256).max` on L2 for example), 0 is a feature missing in Aave right now
35. :construction: not sure about every scenario, for example if we want to completely deprecate a market but the Morpho address has still some supply
36. :x: 
37. :white_check_mark: 
38. :white_check_mark: 
39. :white_check_mark: 
40. :white_check_mark: we can remove the struct entirely
41. :x: too late now IMO
42. :x:
43. :x: 
44. :x: 
45. :x: the type system should prevent anything weird
46. :white_check_mark: already fixed on main (but we not update the submodule without separating prod with other data-structures)
47. :white_check_mark: 
48. :white_check_mark: 
49. :white_check_mark: 
50. :white_check_mark: 
51. :white_check_mark: 
52. :white_check_mark: 
53. referenced in other issues
54. :white_check_mark: with the explanation of pause flags, the document could be called "contingency plan" :sunglasses: 
55. :x: sanity checks are not enough (e.g supply paused on Compound cannot simply be solved by supply paused on Morpho), please refer to the contingency plan :)
56. :x: we don't plan to have p2p on markets with borrow caps. If we do in the future we should assess the risks and implement needed checks
57. :construction: isn't this useful for integrators ? For example they might want to repay max then withdraw max everytime they want to exit a position, whether or not they are actually borrowing
58. :white_check_mark: 
59. :white_check_mark: 
60. :white_check_mark: 
61. referenced in other issues
62. :white_check_mark:
63. :white_check_mark:
64. :white_check_mark: 
65. :white_check_mark: 
66. :white_check_mark: same as 62-65 we should not let Aave revert if remaining liquidity is small
67. :x: we need a document to be ready for this. We can start by restating the non-liquidation theorem that is proved in the yellow
68. :x: not sure how to fix that, and it seems to require some significant initial capital
69. :white_check_mark: I don't see the benefit of having the possibility to set the rewards manager
70. :x: it is not breaking Morpho's specification
71. :x: the same document as the one we write for 67 should apply here
</details>

**Comment by QGarchery** [comment]
Discussion for [issue 34](https://www.notion.so/morpho-labs/34-5710844b99ec4a2db03547c7abc7e31f): **maxSortedUsers has no upper bounds validation and is not the same in Compound/Aave-2**

I don't believe we should do anything regarding this issue. We agree that we should not set a higher limit, and in the same vein having the possibility to set to 0 the max sorted users variable seems interesting. Setting to 0 is prevented in Aave because of an edge case in the implementation that could be fixed in the future, see https://github.com/morpho-dao/morpho-data-structures/issues/80

For Compound we don't have that restriction, and we might want to set MSU to 0 to not have any sorting at all. I feel like the benefit of having a sanity check is small enough that it does not outweight the loss of this feature

**Reply by MerlinEgalite** [reply]
Ok then I agree we can do nothing

**Comment by QGarchery** [comment]
Discussion for [issue 35](https://github.com/spearbit-audits/morpho-novemberAudit/issues/35): **onBehalf argument can be set as the Morpho protocols address**

It seems to be fine for now, but I'm not confident that this will not bite us later. Two examples where this might be tricky and we might forget about it:
- if we want to completely deprecate a market but the Morpho address still has some supply
- if the pool implements a new "mode" (similar to isolation mode) that specifically puts you in a particular mode if you have some assets supplied

**Reply by MerlinEgalite** [reply]
> if the pool implements a new "mode" (similar to isolation mode) that specifically puts you in a particular mode if you have some assets supplied

I'm not sure about this second point. You can still have the behavior even if the onBehalf feature wouldn't exist, no?

**Comment by QGarchery** [comment]
Discussion for [issue 69](https://github.com/spearbit-audits/morpho-novemberAudit/issues/69): **Setting a new rewards manager breaks claiming old rewards**

The rewards manager has some meaningful storage that should not be wiped. It is behind a proxy, and should be updated, so I think that having the ability to set is useless and dangerous. I think we should remove the rewards manager setter in Compound

**Reply by MerlinEgalite** [reply]
remove* you mean. I agree with this.

I'm for totally removing the logic related to rewards in Aave as there is very low likelihood that they add rewards later on the v2 while the v3 is coming.
